pex/pip.py

```diff
@@ -180,7 +195,112 @@ def __init__(
         self.isolated = isolated  # type: bool
 
 
+class _LogAnalyzer(object):
+    @abstractmethod
+    def analyze(self, line):
+        # type: (str) -> Union[bool, str]
+        """Analyze the given log line.
+
+        Returns useful text or else True to continue to analyze the next log line or else False to
+        stop analysis.
```

Comment:
This could maybe be more easily modeled as a coroutine, but it's not clear that the type signature would elucidate its usage enough to avoid the need for the comment.

Reply:
I'm not following the coroutine bit but the types could have been better, fixed those up in 4a615cf.

Reply:
...mm, yea. That was a half baked thought. Was thinking something like [`Coroutine[bool, str, str]`](https://docs.python.org/3/library/typing.html#typing.Coroutine)... but that would definitely be more awkward to use.